### PR TITLE
add 'forbidExplicitDeclaration' option for display-name rule

### PR DESCRIPTION
**Pull request by jakemmarsh**
When using JSX and accepting the transpiler name for `displayName`, it may be preferable to then forbid explicit `displayName` definitions. This adds that option to the rule

### CODE REVIEW

File: docs/rules/display-name.md [hunk 1]
```diff
@@ -85,6 +88,54 @@ function HelloComponent() {
 module.exports = HelloComponent();
 ```
 
+### `forbidExplicitDefinition`
+
+When `true` the rule will not accept any explicitly defined `displayName` property, allowing for the name to be set by the transpiler.
```
Comment (TheSavior) [inline docs/rules/display-name.md]: nit: 

```
to **only** be set by the transpiler.
```

Comment (jakemmarsh) [inline docs/rules/display-name.md]: added

File: lib/rules/display-name.js [hunk 2]
```diff
@@ -191,10 +206,13 @@ module.exports = Components.detect(function(context, components, utils) {
       var list = components.list();
       // Report missing display name for all components
       for (var component in list) {
-        if (!list.hasOwnProperty(component) || list[component].hasDisplayName) {
+        if (forbidExplicitDefinition && list[component].hasDisplayName) {
```
Comment (yannickcr) [inline lib/rules/display-name.js]: The `hasDisplayName` property is `true` if we find an explicit displayName or if the transpiler can find one.

I think we should differentiate the 2 "types" of names and only warn if this is an explicit displayName.

**Comment by yannickcr** [comment]
Hi, and sorry for the review delay.

Thanks for this useful option, I'm totally ok to add it to the `display-name` rule.

But I found a small issue, since the `hasDisplayName` property is also `true` if the transpiler can find a name, this option will warn in all cases, with or without an explicit displayName.

Failing example:
```js
class Hello extends React.Component {
  render() {
    return <div>Hello {this.props.name}</div>;
  }
}
```
Will warn:
```
Explicit display name definition is forbidden
```

**Comment by jakemmarsh** [comment]
No worries about the delay @yannickcr. I'll work on differentiating between the two types of names ASAP

**Comment by ljharb** [comment]
@jakemmarsh altho it’s been awhile, are you still interested in merging this?

**Pull request closed without merging**